source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -552,81 +572,63 @@ void RtFileIo::read(const FilePath& documentPath, const FileSearchPath& searchPa
         readOptions.skipConflictingElements = true;
         readFromXmlFile(document, documentPath, searchPaths, &readOptions);
 
-        CopyOptions copyOptions;
-        copyOptions.skipConflictingElements = true;
-
-        read(document, nullptr, filter);
+        PrvStage* stage = data()->asA<PrvStage>();
+        readDocument(document, stage, filter);
     }
     catch(Exception&)
     {
         return;
     }
 }
 
-void RtFileIo::loadLibraries(const StringVec& libraryPaths, const FileSearchPath& searchPaths)
+void RtFileIo::write(const FilePath& documentPath, const XmlWriteOptions* writeOptions, RtFileIo::WriteFilter filter)
 {
-    // We must create a single document with all dependents since the read
-    // functionality current looks for definitions within the same document.
-    DocumentPtr libraryDoc = createDocument();
-    StringVec libraryNames = MaterialX::loadLibraries(libraryPaths, searchPaths, libraryDoc);
-
     PrvStage* stage = data()->asA<PrvStage>();
-    for (const string& uri : libraryNames)
-    {
-        RtStage libraryStage = RtStage::createNew(RtToken(uri));
-        RtFileIo libStageIo(libraryStage.getObject());
-        RtStage searchStage(getObject());
-        libStageIo.read(libraryDoc, &searchStage,
-            [uri](const ElementPtr &e)
-        {
-            return (e->getActiveSourceUri() == uri);
-        });
-        stage->addReference(libStageIo.data());
-    }
+
+    DocumentPtr document = createDocument();
+    writeDocument(document, stage, filter);
+    writeToXmlFile(document, documentPath, writeOptions);
 }
 
-void RtFileIo::write(DocumentPtr& doc, RtFileIo::WriteFilter filter)
+void RtFileIo::loadLibraries(const StringVec& libraryPaths, const FileSearchPath& searchPaths)
 {
     PrvStage* stage = data()->asA<PrvStage>();
-    writeAttributes(stage, doc);
 
-    for (size_t i = 0; i < stage->numChildren(); ++i)
+    // Load all content into a document.
+    DocumentPtr doc = createDocument();
+    MaterialX::loadLibraries(libraryPaths, searchPaths, doc);
```

Comment:
Again collapsing this into 1 document only avoids the problem of loading in a document after this which uses any dependent nodedef.